Handle session expiration by clearing session when receiving a 401 error

RaghavArora14 opened the pull request:
## Issue Description

Fixes #1933

When a user's session expires or is invalidated on the backend but remains in localStorage, the UI incorrectly shows them as logged in while all API calls fail with 401 errors. This creates:
- Multiple "Unauthorized" toast notifications spamming the screen (one per failed API call)
- UI showing logged-in state while user can't perform authenticated actions
- No automatic recovery - stale session persists until manually cleared

## Changes Applied

### `src/hooks/use-axios.tsx`
- **Toast deduplication**: Added `has401ToastBeenShown` flag to prevent duplicate toasts when multiple API calls fail simultaneously. Flag resets after 6 seconds.
- **Automatic session cleanup**: On 401 detection, immediately call `setSession(null)` to clear stale session from localStorage, updating UI to logged-out state.
- **Conditional toast**: Only show "Unauthorized" toast when a session exists (prevents errors on public pages).

## Before Video


https://github.com/user-attachments/assets/d2d87095-4a0b-4c2e-9c44-f965b537e9dd





## After Video


https://github.com/user-attachments/assets/e132993b-b26c-4285-8cd3-30d95fa9ac41


Discord Thread:
https://discord.com/channels/1233487248129921135/1435373358442090717

seveibar reviewed (changes requested):
A 401 doesnt always mean the seasion expired. Check the jwt via a decode (not a verify) to see if its expired

also tell the user you signed them out via a toast

imrishabh18 commented on src/hooks/use-axios.tsx:
```diff
@@ -31,24 +35,37 @@ export const useAxios = () => {
+          // Reset the flag after a delay so future 401s can show toast again
+          setTimeout(() => {
+            has401ToastBeenShown = false
+          }, 6000)
```
Not sure if this is the correct pattern, cause all the components will be having the same `setTimeout` which might cause race conditions.

seveibar commented on bun.lock:
```diff
@@ -3,6 +3,9 @@
     "": {
       "name": "@tscircuit/fake-snippets",
+      "dependencies": {
+        "jwt-decode": "^4.0.0",
```
Use jose!

RaghavArora14 commented:
https://github.com/user-attachments/assets/f1325164-4bee-4aa4-ba02-d002bf989c1e

RaghavArora14 commented:
rn this does have two toast messages, one for the sign in prompt and one for the error if youd like i can combine these two together

seveibar commented on src/hooks/use-axios.tsx:
```diff
@@ -95,5 +80,5 @@ export const useAxios = () => {
 
     return instance
-  }, [session?.token])
+  }, [session?.token, snippetsBaseApiUrl, setSession, toastLibrary, signIn])
```
Dont add unnecessary proeprties to useEffect (we meep useEffect minimal)

RaghavArora14 commented on src/lib/axios-error-handler.tsx:
```diff
@@ -0,0 +1,115 @@
+export const createAxiosErrorHandler = (
+  session: Store["session"],
+  setSession: (session: Store["session"]) => void,
+  toastLibrary: typeof import("react-hot-toast").default,
```
done, Fixed by importing [toast] directly in the error handler instead of passing it as a parameter, removing unnecessary coupling

RaghavArora14 commented on src/hooks/use-axios.tsx:
```diff
@@ -95,5 +76,5 @@ export const useAxios = () => {
 
     return instance
-  }, [session?.token])
+  }, [session?.token, snippetsBaseApiUrl])
```
oh sorry that's a mistake, you're right should have been removed, fixed it

seveibar reviewed (changes requested):
Not what i meant by the breakup- this is messy and an anti-pattern for hook usage.

1. your error handler create function is too broad, it should be specific to 401 handling
2. You cannot pass handlers like setSession or signIn into a function like that-its a bad inverted control flow pattern.

seveibar commented:
Remember you can create hooks instead of closures-

imrishabh18 commented on src/lib/axios-error-handler.tsx:
Move it to some util folder

RaghavArora14 commented:
Changes applied: 
1. Fixed Critical 401 Bug
Session now ONLY clears if JWT token has expired (checks exp claim)
Does NOT clear session on normal 401 authorization errors
Prevents users being signed out when accessing unauthorized resources
2. Fixed Type Issues
Changed NodeJS.Timeout → number (correct for browser environment)
Used window.setTimeout explicitly
3. Fixed Memory Leaks
Added useEffect cleanup to clear timeout refs on unmount
Added signInPromptTimerRef tracking for proper cleanup
4. Fixed Dependency Array
Changed [session, setSession, signIn] → [session]
Removed unused useSignIn import
5. Renamed Function & File for Clarity
isSessionValid → isSessionExpired (expiration check only, not "validity")
is-session-valid.ts → is-session-expired.ts
Function name now matches file name
6. File Organization
Created src/utils/is-session-expired.ts
Deleted src/lib/axios-error-handler.tsx
Updated imports in use-axios-401-handler.tsx

RaghavArora14 commented on src/hooks/use-axios-401-handler.tsx:
```diff
@@ -0,0 +1,63 @@
+    if (!session?.token) return
+
+    // Always clear the session on 401 (server rejected the token)
+    setSession(null)
```
fixed in [ebb81ac](https://github.com/tscircuit/tscircuit.com/pull/1955/commits/ebb81ac23ad2fc43604f16a260a0dc803b947975)
